### PR TITLE
Make the url generator available on the events

### LINKED ISSUE #286: Make url generator available on `SitemapPopulateEvent`
**Description**  
A very common use case for using this event is to use the url generator to generate URLs. I think it would be a nice quality of life improvement to make the url generator available on the event itself. If of interest, I can make a PR.

**Example**  
```php
$event->getUrlGenerator()->generate(...);
```

### PR DESCRIPTION
**Pull request by kbond**
Fixes #286. Tests pass still so hopefully that means this is BC. Before I finalize, I'd like to ensure this direction is acceptable.

TODO:
- [x] Add tests
- [x] Trigger deprecation warnings
- [x] Document

### CODE REVIEW

File: src/Event/SitemapPopulateEvent.php [hunk 1]
```diff
@@ -39,13 +40,20 @@ class SitemapPopulateEvent extends Event
...
      */
-    public function __construct(UrlContainerInterface $urlContainer, string $section = null)
+    protected $urlGenerator;
+
+    /**
+     * @param UrlContainerInterface      $urlContainer
+     * @param string|null                $section
+     * @param UrlGeneratorInterface|null $urlGenerator
+     */
+    public function __construct(UrlContainerInterface $urlContainer, string $section = null, UrlGeneratorInterface $urlGenerator = null)
```
Comment (kbond) [inline src/Event/SitemapPopulateEvent.php]: Thinking we could trigger a deprecation warning if `$urlGenerator` is null - WDYT?

File: src/Event/SitemapAddUrlEvent.php [hunk 2]
```diff
@@ -49,13 +50,20 @@ class SitemapAddUrlEvent extends Event
...
      */
-    public function __construct(string $route, array $options)
+    protected $urlGenerator;
+
+    /**
+     * @param string                     $route
+     * @param array<string, mixed>       $options
+     * @param UrlGeneratorInterface|null $urlGenerator
+     */
+    public function __construct(string $route, array $options, UrlGeneratorInterface $urlGenerator = null)
```
Comment (kbond) [inline src/Event/SitemapAddUrlEvent.php:63]: I assumed you would want this even to include the `$urlGenerator` as well. If so, we should trigger a deprecation warning here when it is null.

File: src/Service/AbstractGenerator.php [hunk 3]
```diff
@@ -45,20 +46,27 @@ abstract class AbstractGenerator implements UrlContainerInterface
...
 
     /**
-     * @param EventDispatcherInterface $dispatcher
-     * @param int|null                 $itemsBySet
+     * @param EventDispatcherInterface   $dispatcher
+     * @param int|null                   $itemsBySet
+     * @param UrlGeneratorInterface|null $urlGenerator
      */
-    public function __construct(EventDispatcherInterface $dispatcher, int $itemsBySet = null)
+    public function __construct(EventDispatcherInterface $dispatcher, int $itemsBySet = null, UrlGeneratorInterface $urlGenerator = null)
```
Comment (kbond) [inline src/Service/AbstractGenerator.php]: Trigger deprecation when `null` here?

File: src/Event/SitemapAddUrlEvent.php (hunk 2 above)
Comment (yann-eugone) [reply src/Event/SitemapAddUrlEvent.php:63]: Not sure what would be the use case here, but why not.

But yes I think a deprecation should be triggered here if `$urlGenerator` is null.

File: src/Event/SitemapAddUrlEvent.php [hunk 4]
```diff
@@ -123,4 +131,9 @@ public function getOptions(): array
     {
         return $this->options;
     }
+
+    public function getUrlGenerator(): ?UrlGeneratorInterface
```
Comment (yann-eugone) [inline src/Event/SitemapAddUrlEvent.php]: IMO we should have a strict return type, and throw an exception if property is null.

Something that tell the developer what he did wrong.

File: src/Event/SitemapPopulateEvent.php (hunk 1 above)
Comment (yann-eugone) [reply src/Event/SitemapPopulateEvent.php]: Yes, 100% agreed

File: src/Event/SitemapPopulateEvent.php [hunk 5]
```diff
@@ -65,4 +73,9 @@ public function getSection(): ?string
     {
         return $this->section;
     }
+
+    public function getUrlGenerator(): ?UrlGeneratorInterface
```
Comment (yann-eugone) [inline src/Event/SitemapPopulateEvent.php]: Same has above: strict return type, and throw an exception if property is null.

File: src/Service/AbstractGenerator.php (hunk 3 above)
Comment (yann-eugone) [reply src/Service/AbstractGenerator.php]: Yes, here too

**Review by yann-eugone — CHANGES_REQUESTED** [review]
Sorry I'm late (long weekend :))

I think this is a very good start, thank you.

Do you planned to update/add tests in the same PR ?

File: src/Event/SitemapAddUrlEvent.php (hunk 4 above)
Comment (kbond) [reply src/Event/SitemapAddUrlEvent.php]: Perfect, I thought the same thing.

**Comment by kbond** [comment]
> Do you planned to update/add tests in the same PR ?

Yes, that was my plan.

**Comment by kbond** [comment]
Ok, addressed the comments. Let me know if you'd like me to change the deprecation/exception messages.

For the tests, I was thinking the following:
1. Mock/inject the `UrlGeneratorInterface` in the unit tests to remove the self-deprecations
2. Adjust the integration test's `SitemapListener` to use the url generator from the event (instead of injecting)

Let me know if you'd like anything else.

File: src/Event/SitemapAddUrlEvent.php (hunk 2 above)
Comment (yann-eugone) [reply src/Event/SitemapAddUrlEvent.php:63]: Night passed, I'm wondering whether these deprecations at the event level are making sense.

We agree that this is a DX thing ?
A developer that rely on an event the bundle produces will be able to use this feature.
But if he do not want to use this feature and produces is own event, we don't care, do we ?

File: composer.json [hunk 6]
```diff
@@ -17,7 +17,8 @@
         "php": ">=7.1.3",
         "ext-simplexml": "*",
         "symfony/console": "^4.4|^5.0",
-        "symfony/framework-bundle": "^4.4|^5.0"
+        "symfony/framework-bundle": "^4.4|^5.0",
+        "symfony/deprecation-contracts": "^2.4"
```
Comment (yann-eugone) [inline composer.json]: I'm not a big fan of adding a dependency in a minor version.
We usually have issues with dependencies version compatibility : X requires Y 1.x, but Z requires Y 2.x

So far we did not used this library, we were triggering deprecations manually.
Do you think you can go back to that strategy ?

**Comment by yann-eugone** [comment]
>     1. Mock/inject the `UrlGeneratorInterface` in the unit tests to remove the self-deprecations

I know that this bundle has some bad practices in tests, but I'm trying to fix this.
I'm trying to apply the "don't mock what you don't own" approach.
I suggest you inject an implementation of the UrlGeneratorInterface, some actual object from symfony ecosystem, or a test dummy you will create.

>     2. Adjust the integration test's `SitemapListener` to use the url generator from the event (instead of injecting)

:+1:, 100% agree with this

**Review by yann-eugone — CHANGES_REQUESTED** [review]

File: src/Event/SitemapAddUrlEvent.php (hunk 2 above)
Comment (kbond) [reply src/Event/SitemapAddUrlEvent.php:63]: Sure, that's fair. I'll remove.

File: composer.json (hunk 6 above)
Comment (kbond) [reply composer.json]: Sure, I can do this.

**Comment by kbond** [comment]
Ok, I believe this is ready for a final review. I believe the phpstan errors are unrelated.

**Review by yann-eugone — APPROVED** [review]
Thank you for this work !

**Pull request merged**